### DISCUSSION TITLE
This is how you can apply transparency/opacity in individual CTk widgets

### DISCUSSION THREAD

**Discussion by Akascape** [TomSchimansky/CustomTkinter#2214] [Ideas]
Customtkinter widgets are square in shape because it uses tk canvas, the corner_radius and bg_color are camouflaged with the background in such a way that it looks like a rounded widget. But in reality, it is a method just to hide the corners, which doesn't work for overlapping widgets. Hence, it is still an issue with customtkinter.

However, I found the way to fix this. **(ONLY WORKS FOR WINDOWS)**
First, install this special package called [pywinstyles](https://github.com/Akascape/py-window-styles) _(pip install pywinstyles)_
Then simply use this line for the widget: **pywinstyles.set_opacity(_widget_var_, color=_bg_color_)**
Example:
```python
# A basic example of overlapping image and button
import customtkinter
from PIL import Image
import pywinstyles

HEIGHT = 500
WIDTH = 500

app = customtkinter.CTk()
app.title("example")
app.geometry((f"{WIDTH}x{HEIGHT}"))
app.resizable(False, False)

Label1 = customtkinter.CTkLabel(master=app, text="", image=customtkinter.CTkImage(Image.open('image.jpg'), size=(500,500)),
                                width=500, height=500)
Label1.place(x=0, y=0)

Button1 = customtkinter.CTkButton(master=app, width=255, height=172, corner_radius=48,
                                  text='BUTTON', bg_color="#000001") 
Button1.place(x=120, y=57)

pywinstyles.set_opacity(Button1, color="#000001") # just add this line

app.mainloop()
```
![](https://github.com/TomSchimansky/CustomTkinter/assets/89206401/bcddaad8-b465-4c11-a40a-c5887324201a)

### You can even set opacity (partial transparency) in individual widgets like this:
```python
pywinstyles.set_opacity(Button1, value=0.5, color="#000001") # color is not necessary in all cases
```
(_This thing is not available in tkinter by default, exclusively made for tk/ctk_)

![](https://github.com/TomSchimansky/CustomTkinter/assets/89206401/1c9bfce0-8a7b-4d81-9d0f-c9efd592e44a)

## So, this is one more advanced step to take customtkinter to the next level.

**Comment by zain-ziad** [comment]
Trying to use this to make shadows around buttons and frames, but it results in black pixels around the corners. If there's a better way to do this, please let me know. Thanks!

![image](https://github.com/TomSchimansky/CustomTkinter/assets/28985365/7e073e97-b7ae-4192-a236-212ae45b0a67)

Code:
```
import customtkinter as ctk
import pywinstyles

class CustomCTkButton(ctk.CTkFrame):
    def __init__(self, parent, shadow_color='lightgrey', **kwargs):
        super().__init__(parent, fg_color='transparent')
        self.shadow = ctk.CTkFrame(self, fg_color=shadow_color)
        self.button = ctk.CTkButton(self, bg_color="#000001", **kwargs)
        self.button.place(x=0, y=0)
        h = self.button.cget('height')
        w = self.button.cget('width')
        cr = self.button.cget('corner_radius')
        self.shadow.configure(height = h, width = w, corner_radius = cr)
        self.shadow.place(x=0, y=4)
        self.button.lift()
        pywinstyles.set_opacity(self.button, color="#000001")


# Example usage
if __name__ == '__main__':
    app = ctk.CTk()
    custom_button = CustomCTkButton(app, text="Custom", width=200, height=80, corner_radius = 32)
    custom_button.pack()
    app.mainloop()
```

**Reply by Akascape** [reply]
@ZAINheroOFtime This is because tkinter doesn't have antialiasing, you have to change the bg colors of button to white or something similar to fg color of frame in this case so that it matches the background. Instead pass the bg color of button (using cget) to pywinstyles without changing the button bg color. But remember color should not be a tuple.

**Reply by zain-ziad** [reply]
Works great, thank you. Any plans to have this working on Mac or Linux?

**Reply by Akascape** [reply]
@ZAINheroOFtime In linux, there is no transparent option, but only opacity. [(See this issue)](https://stackoverflow.com/questions/75594687/how-to-apply-transparent-background-in-tkinter-window-of-linux-not-alpha)

For linux I guess we have to play with the libx11 library which I really don't know how it works.
Similary on mac, I don't know which library to tweak with ctypes. There is no resource available for that, luckily microsoft has documentation on all their windows internal libraries and functions, which are pretty easy to use with ctypes with windll. Tkinter is also little buggy with mac sometimes.

Bad luck for mac and linux users.

**Reply by Luppyn** [reply]
> Instead pass the bg color of button (using cget) to pywinstyles without changing the button bg color. But remember color should not be a tuple.

I found out there’s a color reduction value applied to the widget. If you take the background’s hex or RGB and subtract the foreground, you get the difference between them. Adding that difference back to the foreground gives the widget the color it was supposed to have all along, but with rounded corners and no artifacts.

**Comment by Bookinheaven** [comment]
pywinstyles.set_opacity(frame1, value=0.5, color="#000001") this method is not working
How to use it with frames?
![xas](https://github.com/TomSchimansky/CustomTkinter/assets/144532387/c61fad24-05a2-49ad-ac1a-d94197b23fb1)

**Comment by feboyfierlyan** [comment]
I LITERALLY HAVE THIS PROBLEM FOR AGES.. doing such things as set the bg_color='transparet', background_corner_colors, even try to manipulate it by giving it another frame and border such stuff but it just doesn't work. Thank you so much for providing this solution!

**Comment by IsEmmanuel** [comment]
En CTk py como hago que la ventana abra maximizada

**Reply by Akascape** [reply]
```python
root.after(10, lambda: root.wm_state("zoomed"))
```

**Comment by MmrMrph** [comment]
I started studying the customtkinter module, but I never joined the community on git. So, I always tried to eliminate the "background" of the widget and I couldn't, so thank you for finding a way around it. Well, I don't like to copy codes but I'll use yours as a base to create my own code to eliminate the "backgrownd" of the widgets!

**Comment by mr-faysal** [comment]
thanks

**Comment by Luppyn** [comment]
Apparently it doesn’t work. I tried to make a decorator using this, but the `set_opacity `method is not assignable to `CTkButton`. I also tried with `CTkLabel `and got the same error. In fact, this method only accepts an integer value, so no CTk widget will work (unless it’s possible to convert it to an int, which I think is unlikely.)

**Reply by Luppyn** [reply]
uh... removing the int type of arg widget in the method I could use correctly!
